### PR TITLE
expression: process boundary value when add generated columns with year type

### PR DESCRIPTION
**Pull request by spongedu**
### What problem does this PR solve? 

Fix issue #9213 .

### What is changed and how it works?
Add a check on builtinCase if the target type is `YEAR`

### Check List 

Tests 

 - Unit test
 - Integration test

Code changes

Side effects

Related changes

 - Need to cherry-pick to the release branch

### CODE REVIEW

File: expression/builtin_cast.go [hunk 1]
```diff
@@ -107,6 +115,53 @@ var (
...
+	if args[0].GetType().Hybrid() || IsBinaryLiteral(args[0]) {
+		sig = &builtinCastIntAsIntSig{bf}
+		sig.setPbCode(tipb.ScalarFuncSig_CastIntAsInt)
+		return sig, nil
+	}
+	argTp := args[0].GetType().EvalType()
+	switch argTp {
+	case types.ETInt:
+		sig = &builtinCastIntAsYearSig{bf, &builtinCastIntAsIntSig{bf}}
+		sig.setPbCode(tipb.ScalarFuncSig_CastIntAsInt)
```
Comment (zz-jason) [inline expression/builtin_cast.go]: it should be `tipb.ScalarFuncSig_CastIntAsYear`?

Comment (spongedu) [reply expression/builtin_cast.go]: @zz-jason we should add function signature like `tipb.ScalarFuncSig_CastIntAsYear` in `tipb`. Before that, I uses `tipb.ScalarFuncSig_CastIntAsInt` instead. 
Maybe we should not set signature before  that's done ?

Comment (spongedu) [reply expression/builtin_cast.go]: @zz-jason I've fixed the pb code. This depends on https://github.com/pingcap/tipb/pull/114

**Comment by spongedu** [comment]
Hi, @XuHuaiyu ,Sorry for delay so long on this issue.  I've take a look how Mysql deals with this.  As is described in [field.cc](https://github.com/mysql/mysql-server/blob/3701bd36bec259fe494c75c9b30c57b01e598297/sql/field.cc#L6349),  method `Field_year::store` casts value of other types into `longlong` and checks data boundary.   As for `Generated Columns`, method [update_generated_read_fields](https://github.com/mysql/mysql-server/blob/3701bd36bec259fe494c75c9b30c57b01e598297/sql/table.cc#L7717) and [update_generated_write_fields](https://github.com/mysql/mysql-server/blob/3701bd36bec259fe494c75c9b30c57b01e598297/sql/table.cc#L7845) evaluate the real value from `generated expression`.

**Comment by spongedu** [comment]
It seems that there's no need to add `CastAsYear` sig, indeed. There's better way to deal with it. I'll close this pr and fix this issue in another way.

**Comment by spongedu** [comment]
In MySQL,   every time a row is read from the storage engine, function `update_generated_read_fields` is called to evaluate the value of generated  columns. There seems no such mechanism in TiDB for now. @XuHuaiyu  @zz-jason

**Comment by spongedu** [comment]
@XuHuaiyu now I manually cast data to year type in `DataSource` operator. This is similar to the MySQL way, I think. PTAL @zz-jason

File: expression/builtin_cast.go [hunk 2]
```diff
@@ -1859,3 +1932,164 @@ func WrapWithCastAsJSON(ctx sessionctx.Context, expr Expression) Expression {
...
 }
+
+func adjustCastYearResult(input int64) int64 {
+	if input > int64(types.MaxYear) || input < int64(types.MinYear) {
+		return 0
+	}
+	return input
+}
+
+type builtinCastIntAsYearSig struct {
```
Comment (zz-jason) [inline expression/builtin_cast.go]: how about only wrap a `builtinCastIntAsIntSig`? like:
```go
type builtinCastIntAsYearSig struct {
	builtinCastIntAsIntSig
}
```

File: expression/builtin_cast.go [hunk 3]
```diff
@@ -1859,3 +1932,164 @@ func WrapWithCastAsJSON(ctx sessionctx.Context, expr Expression) Expression {
...
+	newSig.c.cloneFrom(&b.c.baseBuiltinCastFunc)
+	return newSig
+}
+
+func (b *builtinCastIntAsYearSig) evalInt(row chunk.Row) (res int64, isNull bool, err error) {
+	res, isNull, err = b.c.evalInt(row)
+	if !isNull && err == nil {
+		res = adjustCastYearResult(res)
+	}
+	return
```
Comment (zz-jason) [inline expression/builtin_cast.go]: it's better to `return res, isNull, err` to improve the code readability.

**Comment by zz-jason** [comment]
@spongedu friendly ping, any update?

**Comment by zz-jason** [comment]
I'm going to close this PR due to outdated, fell free to reopen it if you are going to process it again in the future.

**Pull request closed without merging**